shell/platform/android/io/flutter/view/FlutterView.java

```diff
@@ -434,7 +437,90 @@ private int getPointerDeviceTypeForToolType(int toolType) {
+                    state.setUp();
+                    requiredAction.add(new PointerAction(pointerIndex, pointerChange, batched));
+                }
+                break;
+            case kPointerChangeAdd:
+                state = mPointers.get(device);
+                if (state != null) {
+                    state.setLastLocation(dx, dy);
+                } else {
+                    state = new PointerState(device, dx, dy);
+                    mPointers.put(device, state);
+                }
+                requiredAction.add(new PointerAction(pointerIndex, pointerChange, batched));
+                break;
+            case kPointerChangeRemove:
+                state = mPointers.get(device);
+                if (state != null) {
+                    if (state.isDown()) {
+                        requiredAction.add(new PointerAction(pointerIndex, kPointerChangeCancel, batched));
+                    }
+                    requiredAction.add(new PointerAction(pointerIndex, pointerChange, batched));
+                    mPointers.remove(device);
+                }
+                break;
+            case kPointerChangeHover:
+                state = mPointers.get(device);
+                if (state == null) {
+                    state = new PointerState(device, dx, dy);
+                    mPointers.put(device, state);
+                    requiredAction.add(new PointerAction(pointerIndex, kPointerChangeAdd, batched));
+                }
+                requiredAction.add(new PointerAction(pointerIndex, pointerChange, batched));
+                state.setLastLocation(dx, dy);
+                break;
+            case kPointerChangeDown:
+                state = mPointers.get(device);
+                if (state == null) {
+                    state = new PointerState(device, dx, dy);
+                    mPointers.put(device, state);
+                    requiredAction.add(new PointerAction(pointerIndex, kPointerChangeAdd, batched));
+                }
+                state = mPointers.get(device);
+                if (!state.isLastLocation(dx, dy)) {
+                    requiredAction.add(new PointerAction(pointerIndex, kPointerChangeHover, batched));
+                    state.setLastLocation(dx, dy);
+                }
+                requiredAction.add(new PointerAction(pointerIndex, pointerChange, batched));
+                state.setDown();
+                break;
+            case kPointerChangeMove:
+                requiredAction.add(new PointerAction(pointerIndex, pointerChange, batched));
+                state = new PointerState(device, dx, dy);
+                state.setLastLocation(dx, dy);
+                break;
+            default:
+                // unknown action
+                requiredAction.add(new PointerAction(pointerIndex, pointerChange, batched));
+
+        }
+        return requiredAction;
+    }
+
+    private void addToPacket(MotionEvent event, int pointerIndex, int pointerChange,
```

Comment:
Could this be a method on PainterAction that gets passed the packet and adds itself to it?